ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs

```diff
@@ -601,7 +603,13 @@ epochFileParser ImmDbArgs{..} =
     -- It is important that we don't first parse all blocks, storing them all
     -- in memory, and only /then/ extract the information we need.
     decoder' :: forall s. Decoder s (SlotNo, Maybe (HeaderHash blk))
-    decoder' = (\b -> (blockSlot b, immIsEBB b)) <$> immDecodeBlock
```

Comment:
Would it be possible to have a bang on a type smoewhere to make sure that this gets forced? Rather than on the computation of the value?

Reply:
The results of this decoder end up in a list, passing through `readIncrementalOffsets`, `reconstructIndex`, `reconstructSlotOffsets`, etc. Then the list gets reversed before it ends up in a type, i.e. `Index`. Because of this reversing, all of these thunks, each referring to a whole block, end up in memory before we reach `Index`. So unless we use something like `StrictTuple` or replace it with a record with strict fields, we don't have much choice other than force the computation.